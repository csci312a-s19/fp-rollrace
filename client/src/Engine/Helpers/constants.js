/*
 * This file contains all of the constants used throughout the game.
 */

// RENDER_TIMEOUT and STARTING_Y are constants for our constants.
// time between rerenders
const RENDER_TIMEOUT = 20;

// Where the sprite first starts
const STARTING_Y = 547; // currently the same for all maps

// constants object for game initialization
export const CONSTANTS = {
  jump: { STOP: 0, UP: 1, DOWN: 2 }, // Jump state enum for clarity
  UPDATE_INTERVAL: 20, // milliseconds between multiplayer updates
  TOOLBAR_Y: 15, // controls the y positions of the timer and the progress bar
  TOOLBAR_X: 800, // controls the x positions of the timer and the progress bar
  ICON_X: 40, // controls the x position of the pause button and player color icon
  UPDATE_TIMEOUT: 20, // time between motionChange updates
  RENDER_TIMEOUT: RENDER_TIMEOUT, // time between rerenders
  JUMP_SPEED: 0.0013, // acceleration ("gravity") in the game
  JUMP_POWER: 0.7, // jumping velocity
  SCROLL_SPEED: 0.4, // the velocity of the sprite
  SPRITE_SIDE: 100, // size of the invisiable square surrounding the sprite
  PATH_THRESH: 4, // a constant used when finding the end of a path
  TIME_THRESH: RENDER_TIMEOUT, // a constant used in the motionChange update loop
  COUNTDOWN_X: 760, // X position of countdown numbers
  COUNTDOWN_Y: 650, // Y position of countdown numbers
  COUNTDOWN_NUMBERS: ['3', '2', '1', ''], // displayed countdown values
  INITIAL_STATE: {
    countdownIndex: 0, // index of our countdown numbers (where we are in the countdown)
    highscore: '', // current highscore for the current map
    dataSent: false, // indicates if we have sent data (put request) at the end of a game
    wasBooted: false, // indicated whether the game ended due to a boot
    paused: false, // boolean to determine whether or not the game is in the paused state
    endScore: undefined, // score at the completion of the game
    gameover: false, // boolean to determine whether or not the game has ended
    jumpKey: 32, // space bar
    changingKey: false, // boolean to determine whether or not the game is in the key change state
    timerCanStart: false, // prevents timer from starting during countdown
    resetTimer: true, // should we reset the timer after a restart or game start
    y: STARTING_Y, // the current y coordinate of the sprite (upper left corner)
    mapTranslation: 0, // the amount the map is currently translated to the left (interger <= 0)
    hideMenu: false,
    windowHeight: window.innerHeight,
    players: undefined,
    color: `rgb(${Math.random() * 255},${Math.random() * 255},${Math.random() *
      255})` // svg player's color
  },
  INITIAL_VARIABLES: {
    gameStartTime: undefined, // the time the game starts
    x: 200, // this is how far from the left edge of the screen the sprite is
    minY: 1000, // PANIC should loop over all of map or whatever to find this.
    motionChange: undefined, // will take an object of the following form {time: , event: } options for event are block, go, land, and fall
    yStart: STARTING_Y, // a reference point that get's set when the sprite starts moving up or down
    jumpState: 0, // CONSTANTS.jump.STOP === 0 // the current jump state of the game
    jumpStartTime: undefined, // the time a jump begins
    descentStartTime: undefined, // the time a fall begins
    mapTranslationStart: 0, // a reference point that get's set when the sprite starts moving horizontally
    atWall: false, // boolean to determine whether or not the sprite is stopped at a wall
    mapTranslationStartTime: undefined, // the time when the sprite starts moving horizontally
    pauseOffsetStart: undefined, // the time when the pause menu is opened
    timePaused: 0 // the total amount of time the game has spent paused
  }
};
