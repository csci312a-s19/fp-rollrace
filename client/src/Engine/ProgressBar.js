import React from 'react';
import PropTypes from 'prop-types';

const WIDTH = 200; // Set maximum width of progressbar
const HEIGHT = 20; // bar height
const POS_OFFSET = 35;
const SPRITE_OFFSET = 45;

const ProgressBar = props => {
  const { currX, pathLen, spriteColor, x, y } = props;
  /*
    Finds how far the player is as perecntage of the entire path
    Translate that value to the distance from the start the sprite's icon will be on
    the progress bar.
    Currently it only supports single player mode.

  */
  const percentage = currX / pathLen;
  let currentPos = x + WIDTH * percentage;
  if (!currentPos) {
    currentPos = x;
  }

  return (
    <g>
      <rect
        fill={'#ffffff'}
        width={WIDTH}
        height={HEIGHT}
        x={x}
        y={y + POS_OFFSET}
      />

      {/* sprite 1 */}
      <ellipse
        fill={spriteColor}
        cx={currentPos} /*update player one position*/
        cy={y + SPRITE_OFFSET} /* offset */
        rx={2.0}
        ry={9.0}
      />
    </g>
  );
};

ProgressBar.propTypes = {
  y: PropTypes.number.isRequired,
  currX: PropTypes.number,
  x: PropTypes.number.isRequired,
  spriteColor: PropTypes.string.isRequired
};

export default ProgressBar;
