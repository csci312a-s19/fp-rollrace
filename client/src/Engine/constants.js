// Jump state enum for clarity
const RENDER_TIMEOUT = 20;
const STARTING_Y = 547;

export const CONSTANTS = {
  jump: { STOP: 0, UP: 1, DOWN: 2 },
  UPDATE_INTERVAL: 20,
  TOOLBAR_Y: 15,
  TOOLBAR_X: 800,
  ICON_X: 40,
  GAMEOVER_X: 9067,
  UPDATE_TIMEOUT: 20,
  RENDER_TIMEOUT: RENDER_TIMEOUT,
  JUMP_SPEED: 0.0013,
  JUMP_POWER: 0.7,
  SCROLL_SPEED: 0.4,
  SPRITE_SIDE: 100,
  PATH_THRESH: 4,
  TIME_THRESH: RENDER_TIMEOUT,
  INITIAL_STATE: {
    score: '',
    dataSent: false,
    tutorial: false,
    paused: false,
    gameover: false,
    jumpKey: 32, // space bar
    startKey: 115, // s key
    changingKey: false,
    timerCanStart: false,
    y: STARTING_Y,
    mapTranslation: 0,
    hideMenu: false,
    windowHeight: window.innerHeight,
    players: undefined,
    color: `rgb(${Math.random() * 255},${Math.random() * 255},${Math.random() *
      255})`
  },
  INITIAL_VARIABLES: {
    gameStartTime: undefined,
    x: 200,
    minY: 1000, // should loop over all of map or whatever to find this.
    // will take an object of the following form {time: , event: } options for event are block, go, land, and fall
    motionChange: undefined,
    yStart: STARTING_Y,
    jumpState: 0, // CONSTANTS.jump.STOP
    jumpStartTime: undefined,
    descendStartTime: undefined,
    mapTranslationStart: 0,
    atWall: false,
    mapTranslationStartTime: undefined,
    pauseOffsetStart: undefined,
    timePaused: 0
  }
};
