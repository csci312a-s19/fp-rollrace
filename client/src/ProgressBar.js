import React, { Component } from 'react';
import styled from 'styled-components';

//specifies the x coordinate of the rectangle and starting point
const X_POS = 900;

const SpriteOne = styled.ellipse`
  fill: #ff0000;
`;

const SpriteTwo = styled.ellipse`
  fill: #00ff00;
`;

const Bar = styled.rect`
  fill: #280b0b;
`;

class ProgressBar extends Component {
  constructor(props) {
    super(props);
    this.interval = null;
    this.state = {
      player_1_pos: X_POS,
      player_2_pos: X_POS
    };
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  //This is just a placeholder for the purpose of demonstartion
  //the two ellipses go beyond the viewbox but this won't be an issue once
  //we incoroporate it to the game.
  componentDidMount() {
    this.interval = setInterval(() => {
      const player1Pos = this.state.player_1_pos;
      const player2Pos = this.state.player_2_pos;
      this.setState({
        player_1_pos: player1Pos + 5,
        player_2_pos: player2Pos + 10
      });
    }, 1000);
  }

  render() {
    return (
      <g>
        <Bar width={200} height={20} x={X_POS} y={this.props.y} />

        <SpriteOne
          cx={this.state.player_1_pos} /*update player one position*/
          cy={this.props.y + 10} /* offset */
          rx={2.0}
          ry={9.0}
        />

        <SpriteTwo
          cx={this.state.player_2_pos} /*update player two position*/
          cy={this.props.y + 10}
          rx={2.0}
          ry={9.0}
        />
      </g>
    );
  }
}
export default ProgressBar;
