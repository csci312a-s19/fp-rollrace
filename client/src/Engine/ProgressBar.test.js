import React from 'react';
import { mount, shallow } from 'enzyme';
import ProgressBar from './ProgressBar.js';
import { CONSTANTS } from './Helpers/constants.js';

describe('Progress bar snapshot test', () => {
  let comp;
  test('progressBar snapshot', () => {
    comp = shallow(
      <ProgressBar x={CONSTANTS.TOOLBAR_X} y={CONSTANTS.TOOLBAR_Y} />
    );
    expect(comp).toMatchSnapshot();
  });
});

describe('Progress bar components test', () => {
  const x = 800;
  const y = 800;
  const currX = 300;
  const spriteColor = 'red';

  let comp;
  beforeEach(() => {
    comp = mount(
      <ProgressBar x={x} y={y} currX={currX} spriteColor={spriteColor} />
    );
  });

  test('Props testing', () => {
    expect(comp.props().x).toEqual(800);
    expect(comp.props().y).toEqual(800);
    expect(comp.props().currX).toEqual(300);
    expect(comp.props().spriteColor).toEqual('red');
  });

  test('Component has all elements rendered', () => {
    expect(comp.find('rect').exists()).toEqual(true);
    expect(comp.find('ellipse').exists()).toEqual(true);
  });
});
