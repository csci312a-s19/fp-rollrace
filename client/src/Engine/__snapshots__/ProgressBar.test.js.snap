// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress bar snapshot test progressBar snapshot 1`] = `
<g>
  <rect
    fill="#ffffff"
    height={30}
    width={300}
    x={800}
    y={50}
  />
  <ellipse
    cx={800}
    cy={65}
    rx={5}
    ry={15}
  />
</g>
`;
